docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+    - a map from a key to an index
+    - a vector of keys
+    - a vector of values
+- `LookupMap` only has a map from a key to a value. Without a vector of keys, it doesn't have the ability to iterate over keys.
+
+### Performance
+
+`LookupMap` has a better performance and stores less data compared to `UnorderedMap`.
+
+- `UnorderedMap` requires `2` storage reads to get the value and `3` storage writes to insert a new entry.
+- `LookupMap` requires only one storage read to get the value and only one storage write to store it.
+
+### Storage space
+
+`UnorderedMap` requires more storage for an entry compared to a `LookupMap`.
+
+- `UnorderedMap` stores the key twice (once in the first map and once in the vector of keys) and value once. It also has a higher constant for storing the length of vectors and prefixes.
+- `LookupMap` stores key and value once.
+
+## `LazyOption`
+
+It's a type of persistent collection that only stores a single value.
+The goal is to prevent a contract from deserializing the given value until it's needed.
+An example can be a large blob of metadata that is only needed when it's requested in a view call,
+but not needed for the majority of contract operations.
+
+It acts like an `Option` that can either hold a value or not and also requires a unique prefix (a key in this case)
+like other persistent collections.
+
+Compared to other collections, `LazyOption` only allows you to initialize the value during initialization.
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    pub metadata: LazyOption<Metadata>,
+}
+
+#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Metadata {
+    data: String,
+    image: Base64Vec,
+    blobs: Vec<String>,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(metadata: Metadata) -> Self {
+        Self {
+            metadata: LazyOption::new(b"m", Some(metadata)),
+        }
+    }
+
+    pub fn get_metadata(&self) -> Metadata {
+        // `.get()` reads and deserializes the value from the storage. 
+        self.metadata.get().unwrap()
+    }
+}
+```
+
+## Compile smaller binaries
+
+When compiling a contract make sure to pass flag `-C link-arg=-s` to the rust compiler:
+
+```bash
+RUSTFLAGS='-C link-arg=-s' cargo build --target wasm32-unknown-unknown --release
+```
+
+Here is the parameters we use for the most examples in `Cargo.toml`:
+
+```toml
+[profile.release]
+codegen-units = 1
+opt-level = "s"
+lto = true
+debug = false
+panic = "abort"
+overflow-checks = true
+```
+
+You may want to experiment with using `opt-level = "z"` instead of `opt-level = "s"` to see if generates a smaller binary. See more details on this in [The Cargo Book Profiles section](https://doc.rust-lang.org/cargo/reference/profiles.html#opt-level).
```

Comment:
Linking to https://rustwasm.github.io/book/reference/code-size.html#tell-llvm-to-optimize-for-size-instead-of-speed also is probably worth it, as it's much more relevant to what we're using it for.

Reply:
awesome thanks! Decided to move this to a proper section, too https://github.com/near/sdk-docs/pull/37/commits/42d7978f275485d8de3daa9bcdd418bc9f95254d